core/modules/replica/Controller.cc

```diff
@@ -0,0 +1,837 @@
+    }
+
+   /**
+     * Generic method for launching worker service management requests such as suspending,
+     * resyming or inspecting a status of the worker-side replication service.
+     *
+     * @param workerName - the name of a worker node where the service is run
+     * @param onFinish   - a callback function to be called upon completion of the operation
+     */
+    template <class REQUEST_TYPE>
+    static
+    typename REQUEST_TYPE::pointer serviceManagementOperation(
+            Controller::pointer const& controller,
+            std::string const& jobId,
+            std::string const& workerName,
+            typename REQUEST_TYPE::callback_type onFinish,
+            typename Messenger::pointer const& messenger,
+            unsigned int requestExpirationIvalSec) {
+
+        controller->assertIsRunning();
+
+        typename REQUEST_TYPE::pointer request =
+            REQUEST_TYPE::create(
+                controller->_serviceProvider,
+                controller->_io_service,
+                workerName,
+                [controller] (typename REQUEST_TYPE::pointer request) {
+                    controller->finish(request->id());
+                }
+                ,messenger
+            );
+
+        // Register the request (along with its callback) by its unique
+        // identifier in the local registry. Once it's complete it'll
+        // be automatically removed from the Registry.
+
+        (controller->_registry)[request->id()] =
+            std::make_shared<RequestWrapperImpl<REQUEST_TYPE>>(request, onFinish);
+
+        // Initiate the request
+
+        request->start(controller, jobId, requestExpirationIvalSec);
+
+        return request;
+    }
+};
+
+//////////////////////////////////////////////////////////////////////////
+//////////////////////////  ControllerIdentity  //////////////////////////
+//////////////////////////////////////////////////////////////////////////
+
+std::ostream& operator <<(std::ostream& os, ControllerIdentity const& identity) {
+    os  << "ControllerIdentity(id=" << identity.id << ",host=" << identity.host << ",pid=" << identity.pid << ")";
+    return os;
+}
+
+//////////////////////////////////////////////////////////////////
+//////////////////////////  Controller  //////////////////////////
+//////////////////////////////////////////////////////////////////
+
+Controller::pointer Controller::create(ServiceProvider::pointer const& serviceProvider) {
+    return Controller::pointer(new Controller(serviceProvider));
+}
+
+Controller::Controller(ServiceProvider::pointer const& serviceProvider)
+    :   _identity({
+            Generators::uniqueId(),
+            boost::asio::ip::host_name(),
+            getpid()}),
+        _startTime(PerformanceUtils::now()),
+        _serviceProvider(serviceProvider),
+        _numThreads(serviceProvider->config()->controllerThreads()),
+        _io_service(),
+        _work(nullptr),
+        _registry() {
+
+    if (not _numThreads) {
```

Comment:
Fixed. However, if you read the whole section (not just its title) of the Style guid, its last sentence says:
...
if (line). The latter is regarded as such a common practice in C/C++ however that it can be used
...

:-)

Reply:
I think that last line only applies to pointers? It's hard to tell, but I think that's what it means.

Reply:
It's been migrated to an explicit comparison.